File: api/e2e/test/01_create_router_test.go
//go:build e2e

package e2e

import (
	"fmt"
	"net/http"
	"path/filepath"
	"strings"
	"testing"

	"github.com/caraml-dev/turing/api/turing/models"
	"github.com/caraml-dev/turing/api/turing/service"
	"github.com/stretchr/testify/assert"
	"github.com/tidwall/gjson"
)

/*
Steps:
Create a new router with valid config for enricher, ensembler, router.
a. Test GET router immediately > empty config
b. Wait for success response from deployment
c. Test GET router version > status shows "deployed"
d. Test GET router > config section shows version 1, status "deployed"
e. Test cluster that deployments exist
f. Make a request to the router, validate the response.
*/
func TestCreateRouter(t *testing.T) {
	// Create router
	t.Log("Creating router")
	data := makeRouterPayload(
		filepath.Join("testdata", "create_router_nop_logger_proprietary_exp.json.tmpl"),
		globalTestContext)

	withDeployedRouter(t, data,
		func(router *models.Router) {
			t.Log("Testing router endpoint: POST " + router.Endpoint)
			expectedEndpoint := fmt.Sprintf(
				"http://%s-turing-router.%s.%s/v1/predict",
				router.Name,
				globalTestContext.ProjectName,
				globalTestContext.KServiceDomain,
			)
			assert.Equal(t, expectedEndpoint, router.Endpoint)
			withRouterResponse(t,
				http.MethodPost,
				router.Endpoint,
				http.Header{
					"Content-Type":  {"application/json"},
					"X-Mirror-Body": {"true"},
				},
				`{"client": {"id": 4}}`,
				func(response *http.Response, responsePayload []byte) {
					assert.Equal(t, http.StatusOK, response.StatusCode,
						"Unexpected response (code %d): %s",
						response.StatusCode, string(responsePayload))
					actualResponse := gjson.GetBytes(responsePayload, "response").String()
					expectedResponse := `{
						"experiment": {
							"configuration": {
								"foo":"bar",
								"route_name":"treatment-a"
							}
						},
						"route_responses": [
							{
								"data": {
									"version": "control"
								},
								"is_default": false,
								"route": "control"
							}
					  	]
					}`
					assert.JSONEq(t, expectedResponse, actualResponse)
				})

			batchEndpoint := strings.Replace(router.Endpoint, "/predict", "/batch_predict", -1)
			t.Log("Testing router batch endpoint: POST " + batchEndpoint)
			withRouterResponse(t,
				http.MethodPost,
				batchEndpoint,
				http.Header{
					"Content-Type":  {"application/json"},
					"X-Mirror-Body": {"true"},
				},
				`[{"client": {"id": 4}}, {"client": {"id": 7}}]`,
				func(response *http.Response, responsePayload []byte) {
					assert.Equal(t, http.StatusOK, response.StatusCode,
						"Unexpected response (code %d): %s",
						response.StatusCode, string(responsePayload))
					t.Logf("Response Payload:\n%s", string(responsePayload))
					expectedResponse := `[
						{
							"code": 200,
							"data": {
								"request": {
									"client": {
										"id": 4
									}
								},
								"response": {
									"experiment": {
										"configuration": {
											"foo": "bar",
											"route_name":"treatment-a"
										}
									},
									"route_responses": [
										{
											"data": {
												"version": "control"
											},
											"is_default": false,
											"route": "control"
										}
									]
								}
							}
						},
						{
							"code": 200,
							"data": {
								"request": {
									"client": {
										"id": 7
									}
								},
								"response": {
									"experiment": {
										"configuration": {
											"bar": "baz",
											"route_name":"control"
										}
									},
									"route_responses": [
										{
											"data": {
												"version": "control"
											},
											"is_default": false,
											"route": "control"
										}
									]
								}
							}
						}
					]`
					assert.JSONEq(t, expectedResponse, string(responsePayload))
				})

			t.Log("Test endpoints for router logs")
			baseURL, projectID, routerID := globalTestContext.APIBasePath, globalTestContext.ProjectID, router.ID
			url := fmt.Sprintf("%s/projects/%d/routers/%d/logs", baseURL, projectID, routerID)
			componentTypes := []string{"router", "ensembler", "enricher"}
			var podLogs []service.PodLog

			for _, c := range componentTypes {
				queryString := ""
				if c != "" {
					queryString = "?component_type=" + c
				}
				t.Log("GET", url+queryString)
				resp, err := http.Get(url + queryString)
				assert.NoError(t, err)
				assert.Equal(t, http.StatusOK, resp.StatusCode)
				podLogs = getPodLogs(t, resp)
				assert.Greater(t, len(podLogs), 0)
			}
		},
		nil,
	)
}

File: sdk/e2e/01_create_router_test.py
import os
import logging

import requests
import turing
import turing.batch
import turing.batch.config
import turing.router.config.router_config

from turing.router.config.route import Route
from turing.router.config.experiment_config import ExperimentConfig
from turing.router.config.resource_request import ResourceRequest
from turing.router.config.log_config import LogConfig, ResultLoggerType
from turing.router.config.enricher import Enricher
from turing.router.config.common.env_var import EnvVar
from turing.router.config.router_ensembler_config import DockerRouterEnsemblerConfig
from turing.router.config.router_config import RouterConfig, Protocol
from turing.router.config.router_version import RouterStatus


def test_create_router():
    # assert 0 routers are present before creating a new router
    assert len(turing.Router.list()) == 0

    # set up route
    routes = [
        Route(
            id="control",
            endpoint=f'{os.getenv("MOCKSERVER_HTTP_ENDPOINT")}/control',
            timeout="5s",
        )
    ]

    # set up experiment config
    experiment_config = ExperimentConfig(
        type="proprietary",
        config={
            "client": {"id": "1", "username": "test", "passkey": "test"},
            "experiments": [{"id": "001", "name": "exp_1"}],
            "variables": {
                "experiment_variables": {
                    "001": [{"name": "client_id", "type": "unit", "required": True}]
                },
                "config": [
                    {
                        "name": "client_id",
                        "required": True,
                        "field": "client.id",
                        "field_source": "payload",
                    }
                ],
            },
        },
    )

    # set up resource request config for the router
    resource_request = ResourceRequest(
        min_replica=1, max_replica=1, cpu_request="100m", memory_request="250Mi"
    )

    # set up log config for the router
    log_config = LogConfig(result_logger_type=ResultLoggerType.NOP)

    # set up enricher for the router
    enricher = Enricher(
        image=os.getenv("TEST_ECHO_IMAGE"),
        resource_request=ResourceRequest(
            min_replica=1, max_replica=1, cpu_request="100m", memory_request="1Gi"
        ),
        endpoint="anything",
        timeout="2s",
        port=80,
        env=[EnvVar(name="TEST_ENV", value="enricher")],
    )

    # set up ensembler for the router
    ensembler = DockerRouterEnsemblerConfig(
        image=os.getenv("TEST_ECHO_IMAGE"),
        resource_request=ResourceRequest(
            min_replica=2, max_replica=2, cpu_request="100m", memory_request="256Mi"
        ),
        endpoint="anything",
        timeout="3s",
        port=80,
        env=[EnvVar(name="TEST_ENV", value="ensembler")],
    )

    # create the RouterConfig instance
    router_config = RouterConfig(
        environment_name=os.getenv("MODEL_CLUSTER_NAME"),
        name=f'e2e-sdk-experiment-{os.getenv("TEST_ID")}',
        routes=routes,
        rules=[],
        experiment_engine=experiment_config,
        resource_request=resource_request,
        timeout="5s",
        log_config=log_config,
        enricher=enricher,
        ensembler=ensembler,
    )

    # create a router using the RouterConfig object
    router = turing.Router.create(router_config)
    assert router.status == RouterStatus.PENDING

    # wait for the router to get deployed
    try:
        router.wait_for_status(RouterStatus.DEPLOYED)
    except TimeoutError:
        raise Exception(
            f"Turing API is taking too long for router {router.id} to get deployed."
        )
    assert router.status == RouterStatus.DEPLOYED

    # get router with id 1
    retrieved_router = turing.Router.get(1)
    assert retrieved_router.version == 1
    assert retrieved_router.status == RouterStatus.DEPLOYED
    assert retrieved_router.config.protocol == Protocol.HTTP
    assert (
        retrieved_router.endpoint
        == f'http://{retrieved_router.name}-turing-router.{os.getenv("PROJECT_NAME")}.{os.getenv("KSERVICE_DOMAIN")}/v1/predict'
    )

    # get router version with id 1
    router_version_1 = retrieved_router.get_version(1)
    assert router_version_1.status == RouterStatus.DEPLOYED

    # post single request to turing router
    logging.info("Testing router endpoint...")
    response = requests.post(
        url=router.endpoint,
        headers={
            "Content-Type": "application/json",
            "X-Mirror-Body": "true",
        },
        json={"client": {"id": 4}},
    )
    assert response.status_code == 200
    expected_response = {
        "experiment": {
            "configuration": {"foo": "foo", "route_name": "control"},
        },
        "route_responses": [
            {"data": {"version": "control"}, "is_default": False, "route": "control"}
        ],
    }
    assert response.json()["response"] == expected_response

    # post batch request to turing router
    logging.info("Testing router batch endpoint...")
    response = requests.post(
        url=router.endpoint.replace("/predict", "/batch_predict"),
        headers={
            "Content-Type": "application/json",
            "X-Mirror-Body": "true",
        },
        json=[{"client": {"id": 4}}, {"client": {"id": 7}}],
    )
    assert response.status_code == 200
    expected_response = [
        {
            "code": 200,
            "data": {
                "request": {"client": {"id": 4}},
                "response": {
                    "experiment": {
                        "configuration": {"foo": "foo", "route_name": "control"},
                    },
                    "route_responses": [
                        {
                            "data": {"version": "control"},
                            "is_default": False,
                            "route": "control",
                        }
                    ],
                },
            },
        },
        {
            "code": 200,
            "data": {
                "request": {"client": {"id": 7}},
                "response": {
                    "experiment": {
                        "configuration": {"bar": "bar", "route_name": "treatment-a"},
                    },
                    "route_responses": [
                        {
                            "data": {"version": "control"},
                            "is_default": False,
                            "route": "control",
                        }
                    ],
                },
            },
        },
    ]
    assert response.json() == expected_response

    # test endpoint for router logs
    logging.info("Testing endpoint for router logs...")
    router_component_types = ["router", "ensembler", "enricher"]
    base_url = f'{os.getenv("API_BASE_PATH")}/projects/{os.getenv("PROJECT_ID")}/routers/{router.id}/logs'
    for component in router_component_types:
        response = requests.get(
            url=base_url + "?component_type=" + component,
        )
    assert response.status_code == 200
    assert len(response.content) > 0
